### DISCUSSION TITLE
Optional field specific validations

### DISCUSSION THREAD

**Discussion by MikeSniegovski** [colinhacks/zod#974] [Q&A] [answered]
Im generating schema from api data in runtime for validating in Formik.
Is there any possibility to add for egzample **max** to z.number only if it is not null on my config object?
I have no idea how to achive this.

My config object interface

```ts
interface ISurveyQuestion {
    questionId: number;
    title: string;
    subtitle: string | null;
    placeholder: string | null;
    description: string | null;
    imgUrl: string | null;
    hint: string | null;
    unit: string | null;
    inputType: "number" | "checkbox" | "radio" | "select" | "date";
    min: number | null;
    max: number | null;
    options: [ISurveyOption] | null;
}
```

**Comment by christianhg** [comment]
Union types are powerful for this sort of thing. Perhaps this helps you?

```ts
import { z } from "zod";

const Question = z.object({
  max: z.union([z.number().max(42), z.null()])
});

console.log(Question.parse({ max: null })); // OK 
console.log(Question.parse({ max: 30 })); // OK
console.log(Question.parse({ max: 43 })); // Error
```

**Comment by MikeSniegovski** [comment]
Thanks for quick repay, but I'm interested more i something like:

```ts
let numberShema = z.number({ required_error: "survey.validation.required" })

if( question.max) {
    numberShema.max( question.max, { message: "survey.validation.toSmall" })
}
```

The idea here is to add additional validations to field only if it exists in config object.

**Comment by fergusmeiklejohn** [comment]
We have a similar problem. With Remix we need to validate form submissions. So a form might have an email field that is optional but if the user enters their email we need to validate it's an email.
We're having trouble achieving this with Zod: `z.string().optional()` works but `z.string().email().optional()` returns invalid on an empty field `z.union([z.string().email(), z.string().optional()])` returns valid if the user enters an invalid email address.

**Comment by scotttrinh** [comment]
@fergusmeiklejohn 

I think you just need to tweak the union slightly:

[CodeSandbox](https://codesandbox.io/s/relaxed-sea-7xvzm?file=/src/index.ts)

```ts
z.union([z.string().email(), z.undefined()])
```

Basically "it's either an email string OR undefined" rather than "it's either an email string OR a string OR undefined".

---

@MikeSniegovski 

I'm not 100% sure what you're asking for or what a config object is in the context of your question. Can you provide more code (ideally a run-able example like [CodeSandbox](https://codesandbox.io/s/zod-template-r67dv))

**Comment by jgb-solutions** [comment]
Any updates on this? @scotttrinh I tried your example but it doesn't work. So basically I have an optional field that needs to be a certain length of characters when not empty. With `z.undefined()` the fields gets required. I'm using React Hook Form by the way.

**Reply by imaksp** [reply]
See optional string validation section here:
https://github.com/colinhacks/zod#unions

**Comment by JacobWeisenburger** [comment, marked as answer]
> Thanks for quick repay, but I'm interested more i something like:
> 
> ```ts
> let numberShema = z.number({ required_error: "survey.validation.required" })
> 
> if( question.max) {
>     numberShema.max( question.max, { message: "survey.validation.toSmall" })
> }
> ```
> 
> The idea here is to add additional validations to field only if it exists in config object.

@MikeSniegovski 
Is this what you are looking for?

```ts
function makeNumberSchema ( { max }: { max?: number } = {} ) {
    const baseSchema = z.number( { required_error: 'survey.validation.required' } )
    return max != null
        ? baseSchema.max( max, { message: 'survey.validation.toSmall' } )
        : baseSchema
}

console.log( makeNumberSchema().parse( 15 ) ) // 15

const result = makeNumberSchema( { max: 10 } ).safeParse( 15 )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: 'too_big',
//         maximum: 10,
//         type: 'number',
//         inclusive: true,
//         exact: false,
//         message: 'survey.validation.toSmall',
//         path: []
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
